modules/afamqp/afamqp.c

```diff
@@ -451,24 +467,31 @@ afamqp_dd_login(AMQPDestDriver *self)
   switch (self->auth_method)
     {
     case AMQP_SASL_METHOD_PLAIN:
-      ret = amqp_login(self->conn, self->vhost, self->max_channel, self->frame_size, 0,
-                       self->auth_method, self->user, self->password);
+      ret = amqp_login(self->conn, self->vhost, self->max_channel, self->frame_size,
+                       self->offered_heartbeat, self->auth_method, self->user, self->password);
       break;
 
+
     case AMQP_SASL_METHOD_EXTERNAL:
       // The identity is generally not needed for external authentication, but must be set to
       // something non-empty otherwise the API call will fail, so just default it to something
       // meaningless
       identity = (self->user && self->user[0] != '\0') ? self->user : ".";
-      ret = amqp_login(self->conn, self->vhost, self->max_channel, self->frame_size, 0,
-                       self->auth_method, identity);
+      ret = amqp_login(self->conn, self->vhost, self->max_channel, self->frame_size,
+                       self->offered_heartbeat, self->auth_method, identity);
       break;
 
     default:
       g_assert_not_reached();
       return FALSE;
     }
 
+  self->heartbeat = amqp_get_heartbeat(self->conn);
+
+  msg_debug("Amqp heartbeat negotiation",
```

Comment:
This is for the documentation:
As I read the docs, heartbeat can only be turned off, if BOTH sides has a heartbeat value of 0.
https://www.rabbitmq.com/heartbeats.html
> Heartbeats can be disabled by setting the timeout interval to 0 on both client and server ends.

On code side we have this debug message, so there is nothing more to do, it's just worth to mention in the docs.

Reply:
**UPDATE**: I've tried this out and this statement seems invalid.
What I've concluded is that generally the negotiated heartbeat interval will be the smaller configured one.
About disabling heartbeating: Client controls enable/disable, therefore even if the server side has an explicit 0 for heartbeat, a client can override this. If the client has a 0, it means disable heartbeat.
(tested with RabbitMQ 3.6.10 server and client 0.8.0 versions).